### PR TITLE
Register qualified args

### PR DESCRIPTION
**Pull request by nicky-johnstone**
Tcl
----
- [x] Does each proc have a #| comment to describe what it does ?
- [x] Are there comments throughout the code ?
- [x] Is each proc unit testable ?
- [x] Are the proc names well chosen ?
- [x] Are variable names well chosen ?
- [x] List what testing has been done.
  - Used the following test pages with varying input
  ```tcl
  register GET /test1.html {form.foo} {
      set list [list]
      lappend list "FOO: $foo"
      lappend list "FORM.FOO: ${form.foo}"
      lappend list "LOCAL VARS: [info locals]"
  
      return_html [html_list $list]
  }
  
  register GET /test2.html {form.foo form.bar} {
      lappend list "FOO: $foo"
      lappend list "FORM.FOO: ${form.foo}"
      lappend list "BAR: $bar"
      lappend list "FORM.BAR: ${form.bar}"
      lappend list "LOCAL VARS: [info locals]"
  
      return_html [html_list $list]
  }
  
  register GET /test3.html {form.foo form.bar baz} {
      lappend list "FOO: $foo"
      lappend list "FORM.FOO: ${form.foo}"
      lappend list "BAR: $bar"
      lappend list "FORM.BAR: ${form.bar}"
      lappend list "BAZ: $baz"
      lappend list "LOCAL VARS: [info locals]"
  
      return_html [html_list $list]
  }
  
  register GET /test4.html {{form.foo THIS} form.bar form.baz other.baz} {
      lappend list "FOO: $foo"
      lappend list "FORM.FOO: ${form.foo}"
      lappend list "BAR: $bar"
      lappend list "FORM.BAR: ${form.bar}"
      lappend list "FORM.BAZ: ${form.baz}"
      lappend list "OTHER.BAZ: ${other.baz}"
      lappend list "LOCAL VARS: [info locals]"
      
      return_html [html_list $list]
  }
  
  register GET /test5.html {form.foo form.bar {form.baz A} other.baz} {
      lappend list "FOO: $foo"
      lappend list "FORM.FOO: ${form.foo}"
      lappend list "BAR: $bar"
      lappend list "FORM.BAR: ${form.bar}"
      lappend list "FORM.BAZ: ${form.baz}"
      lappend list "OTHER.BAZ: ${other.baz}"
      lappend list "LOCAL VARS: [info locals]"
      
      return_html [html_list $list]
  }
  
  register GET /test6.html {form.baz other.baz baz} {
      lappend list "FORM.BAZ: ${form.baz}"
      lappend list "OTHER.BAZ: ${other.baz}"
      lappend list "BAZ: $baz"
      lappend list "LOCAL VARS: [info locals]"
  
      return_html [html_list $list]
  }
  ```
  - Tested in MLA ERP with new sales order by removing the arg mappings i.e. `set customer_code ${customer.customer_code}`

### CODE REVIEW

File: tcl/util.tcl [hunk 1]
```diff
@@ -624,3 +623,31 @@ proc qc::glob_recursive {args} {
     # Call glob and return the results
     return [glob {*}$switches {*}$options -- {*}$patterns]
 }
+
+proc qc::qualified_args2unqualified {} {
```
Comment (nicky-johnstone) [inline tcl/util.tcl]: Struggled with a decent name here.

Some other ideas:
* `qc::args_map_unqualified`
* `qc::args2unqualified_vars`

**Comment by bvw** [comment]
Code snippets to discuss on Monday

```tcl
proc shortname {name} {if { [regexp {^([^.]+)\.([^.]+)$} $name -> table column] } {return $column} else { return $name }} 
proc map {lambda list} {
    set result {}
    foreach item $list {
        lappend result [apply $lambda $item]
    }
    return $result
}

proc unambiguous {args} {
    set unambiguous {}
    set args [map {x {shortname $x}} $args]
    foreach item $args { 
        if { [llength [lsearch -all $args $item]]==1 } {
            lappend unambiguous $item
        }
    }
    return $unambiguous
}

```

**Comment by nicky-johnstone** [comment]
Updated.

File: tcl/util.tcl [hunk 2]
```diff
@@ -624,3 +623,40 @@ proc qc::glob_recursive {args} {
...
+    #|      Caller's args: { table.foo other.foo }
+    #|      -> No new variables set in caller because "foo" would be ambiguous.
+    set caller_args [info args [dict get [info frame -1] proc]]
+    foreach item [qc::unambiguous {*}$caller_args] {
+        upvar 1 $item value
+        qc::upset 1 [qc::shortname $item] $value
+    }
+}
+
+proc qc::shortname {name} {
```
Comment (bvw) [inline tcl/util.tcl]: I think we better make the names of these procs less special `arg_shortname` `args_unambiguous` or something similar

**Comment by nicky-johnstone** [comment]
Updated. Also renamed `qc::qualified_args2unqualified` to be consistent with the rest.

**Pull request merged**